Add the change passphrase feature

VandalByte opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

### Description
This PR intends to add the much requested and discussed functionality to change passphrase in vorta UI.

### Related Issue
Closes #303 

### Motivation and Context
This change adds the ability to change the passphrase directly within the Vorta backup client through a user-friendly interface. Previously, users had to rely on manual borg command-line operations to update their passphrase. This simplifies the process, improving usability by allowing users to easily update their passphrase without leaving the Vorta interface.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Manual testing and unit testing.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [CONTRIBUTING](https://vorta.borgbase.com/contributing/) guide.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.


*I provide my contribution under the terms of the [license](./../LICENSE.txt) of this repository and I affirm the [Developer Certificate of Origin][dco].*

[dco]: https://developercertificate.org/

<!--
This template is sourced from the awesome https://github.com/TalAter/open-source-templates
-->

VandalByte commented:
This is the design I thought to got with, I have considered a way to logically group this in a new drop down button.
- This button serves more like a utility button for the selected repo.
- Currently, the 'Change Passphrase' option is being added, but the dropdown can be easily expanded to include other options in the future (e.g., 'Change Repo Name' feature).
- I may adjust the icon based on feedback.

Any suggestions regarding the UI design are always welcome!

![Screenshot 2025-04-01 003733](https://github.com/user-attachments/assets/50bcd589-8d82-4ac8-bdac-6a83d9bbaddc)

m3nu commented:
How about moving more lesser-used features (like unlink) into the submenu?

and I'd use 3 dots as the submenu icon. I feel it's more widely used. The cog icon usually means _Settings_.

VandalByte commented:
> How about moving more lesser-used features (like unlink) into the submenu?

I can do that.

> and I'd use 3 dots as the submenu icon. I feel it's more widely used. The cog icon usually means Settings.

Yeah, so I was actually stuck with these two, I feel like the horizontal one looks a bit better. But if we want to go with vertical one then there's already a 3 dot [present](https://github.com/borgbase/vorta/blob/master/src/vorta/assets/icons/ellipsis-v.svg) we can reuse that icon.

![Screenshot 2025-04-03 202046](https://github.com/user-attachments/assets/f1b78d1f-7f5c-4827-98ab-adfd41faa133)

The ellipse-vertical
![image](https://github.com/user-attachments/assets/9bd53f5a-b008-436f-b65a-3fb148cf2c9d)

VandalByte commented:
I have moved the `unlink`, should I keep the icon or remove it?

![image](https://github.com/user-attachments/assets/111250d1-6dd7-4609-85fb-a4287ae4c640)

m3nu commented:
I would either do a button for both or none.

VandalByte commented:
@m3nu I'd like your opinion on a few things. The current implementation changes the passphrase for the currently selected repo (same behavior as how _unlink_ works).

1. At the moment, I'm not validating the current passphrase (i.e., checking whether the old passphrase entered is correct before applying the new passphrase). I'm just directly setting the new one.
    - I felt this might be fine, since from what I saw, vorta when adding an existing repo takes the passphrase from the keyring and applies it, rather than asking the user to enter and confirm it against the keyring value before adding the repo.

2. For the confirmation message, ie, when passphrase was successfully changed, i have two methods in mind:
    - First, show a message right below the `errortext` validation, green if successful red if failed.
    - Second, show a dialog with an "OK" button containing the message.

I am thinking of going with the 1st option, to avoid the need for too many dialogs popping up.

This is the screenshot of the currently implemented UI:
![Screenshot 2025-04-12 110937](https://github.com/user-attachments/assets/09c6ba0b-154a-4fe0-9737-7eac0d8361fd)

m3nu commented:
> I'm just directly setting the new one.

Can you change the passphrase without knowing the existing one? I doubt it.

Still better to run only one command in total for this and throw an error if it fails for any reason (issue with connecting to repo or wrong passphrase or anything else).

> Second, show a dialog with an "OK" button containing the message.

This one is probably better because the user isn't left with an open dialog.

VandalByte commented:
https://github.com/user-attachments/assets/bb40da52-30a6-4bd6-bf61-7629932beb51

VandalByte commented:
Added passphrase change only for encryption type = repokey, as mentioned in the [comment](https://github.com/borgbase/vorta/issues/303#issuecomment-1472556103).

VandalByte commented:
The UI part I made some adjustments for the bottom label property and overall widget layout was changed from `grid` to `vertical`. The input fields are inside a `form layout` idk why it's showing up misaligned (labels would be left aligned) in your screenshot. I also noticed you are missing the widget title "Change Passphrase" and the close button at the top. This is how it looks for me (also resizes flawlessly) in MX Linux XFCE. 

![image](https://github.com/user-attachments/assets/40fb5fdb-701a-45bd-a430-626c216c3eaf)

> It won't update the passphrase in the keychain after changing it. As a result it keeps using the old one and everything fails after changing the passphrase.

Now about the error, could you tell me how to reproduce this? because I couldn't. 

These are the steps I tried:
- I changed passphrase for a local repo and changed it again for the same repo. 
- I changed the passphrase quit vorta and started vorta again and started a manual backup.

Is this maybe a remote repo specific issue? Could you test it out with a local repo and let me know if it works?

m3nu commented:
It's not due to the remote repo. I check the macOS keychain and it still had the old password after the change. After changing the password manually, it worked again.

So it might be macOS-specific. Could also be with keychain permissions.

VandalByte commented:
I might have missed the part to update the keyring since it worked properly in linux during testing (for some reason). I have updated the PR, can you verify?

m3nu commented:
Nope. Password is still unchanged in keyring.

Quite possible that our current macos code doesn't allow overwriting an existing item. The method we use is `SecKeychainAddGenericPassword`.

Does it work on Linux? How did it work on Linux without this code?

I guess we need to add a `delete_password()` method for all `keyring` classes to make this work. The only one that works for sure is our internal `db.py` class, which already uses `get_or_create()`.

m3nu commented:
OK. Will test.

One other thing I had noticed, but forgot to add: The icon for this dialog could be set to type warning or confirmation or something. It may use some default right now. Just whatever we use in other places.

<img width="333" alt="Screenshot 2025-07-01 at 11 18 46" src="https://github.com/user-attachments/assets/60a65aa8-7488-4b4c-afdc-40d09c08986f" />

VandalByte commented:
I've added the appropriate info and warning icon properties to the message boxes. The icons should now display correctly.

m3nu commented on src/vorta/views/repo_tab.py:
```diff
@@ -328,3 +340,38 @@ def copy_URL_action(self):
+
+    def repo_change_passphrase_action(self):
+        if not self.profile().repo.encryption.startswith('repokey'):
+            msg = QMessageBox()
```
Ignore this. Would be a separate PR, since we use many messageboxes just in this one file. Refactoring them will be too much.

VandalByte commented:
Recent changes include:
- Made the `NOTE` label expanding horizontally and added word wrap.
- Added form property to make the LineEdits expand (`FieldGrowthPolicy` as `ExpandingFieldsGrow`).
- Minor changes like margin and alignment.

The changes has been tested in MX Linux 22 (XFCE).

m3nu commented:
This new feature works as expected locally and could be merged right away.

I would just add 2 things (which could be in a separate PR, if this one is getting too big or blocking something):

- At least a simple unit test to make sure this keeps working after future changes.
- I would show the repo URL in the dialog somewhere so it's always clear _which_ passphrase is updated. (given how important this is to the user)

VandalByte commented:
All the requested changes have been included. This should be good to go.

m3nu commented:
Great! Will test one more time and then merge.
